File: src/main/java/leetcode/lists/offer/Offer42.java
package leetcode.lists.offer;

/**
 * @author jingxinwu
 * @date 2021-11-22 10:45 下午
 */
public class Offer42 {


    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        int res = Integer.MIN_VALUE;

        for (int i = 1; i < nums.length; i++) {
            dp[i] = dp[i] + Math.max(nums[i - 1], 0);
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}

File: src/main/go/leetcode/lists/offer/Offer42.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 23:37
 */

func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	max := nums[0]
	for i := 1; i < len(nums); i++ {
		//dp[i] = Max(0, dp[i-1]) + nums[i] same ？
		dp[i] = Max(dp[i-1]+nums[i], nums[i])
		max = Max(max, dp[i])
	}
	return max
}
